pkg/trace/writer/newwriter/sender.go

```diff
@@ -0,0 +1,382 @@
+package writer
+
+import (
+	"bytes"
+	"container/list"
+	"errors"
+	"fmt"
+	"math"
+	"math/rand"
+	"net/http"
+	"net/url"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/trace/info"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// eventType specifies an event which occurred in the sender.
+type eventType int
+
+const (
+	// eventTypeRetry specifies that a send failed with a retriable error (5xx).
+	eventTypeRetry eventType = iota
+	// eventTypeFlushed specifies that a list of one or more payloads was flushed.
+	eventTypeFlushed
+	// eventTypeSent specifies that a single payload was successfully sent.
+	eventTypeSent
+	// eventTypeFailed specifies that a payload failed to send and data was lost.
+	eventTypeFailed
+	// eventTypeDropped specifies that a payload had to be dropped to make room
+	// in the queue.
+	eventTypeDropped
+)
+
+var eventTypeStrings = map[eventType]string{
+	eventTypeRetry:   "eventTypeRetry",
+	eventTypeFlushed: "eventTypeFlushed",
+	eventTypeSent:    "eventTypeSent",
+	eventTypeFailed:  "eventTypeFailed",
+	eventTypeDropped: "eventTypeDropped",
+}
+
+// String implements fmt.Stringer.
+func (t eventType) String() string { return eventTypeStrings[t] }
+
+// eventData represents information about a sender event. Not all fields apply
+// to all events.
+type eventData struct {
+	// host specifies the host which the sender is sending to.
+	host string
+	// bytes represents the number of bytes affected by this event. It is
+	// not known for eventTypeFlushed.
+	bytes int
+	// count specfies the number of payloads that this events refers to.
+	count int
+	// duration specifies the time it took to complete this event. It
+	// is set for eventTypeFlushed, eventTypeSent, eventTypeRetry and
+	// eventTypeFailed.
+	duration time.Duration
+	// err specifies the error that may have occurred on events like
+	// eventTypeRetry and eventTypeFailed.
+	err error
+}
+
+// eventRecorder implementations are able to take note of events happening in
+// the sender.
+type eventRecorder interface {
+	// recordEvent notifies that event t has happened, passing details about
+	// the event in data.
+	recordEvent(t eventType, data *eventData)
+}
+
+// maxQueueSize specifies the maximum allowed queue size. If it is surpassed, older
+// items are dropped to make room for new ones.
+var maxQueueSize = 64 * 1024 * 1024 // 64MB; replaced in tests
```

Comment:
Yeah this won't really be a problem under normal operation. But whenever there's some network outage it directly affects the amount of data that will be lost.

I'm a bit itchy about leaving it hardcoded as 64MB for a particular org might represent 5 minutes of data and for another it might represent 30 seconds. OTOH, we could definitely delay the exposure of this config knob until requested. Making it a fraction of `max_memory` could also work I guess.

This is more of a "product" decision than technical at the end.

Reply:
Ok, let's leave it open and get more input from product people later whether we want to:

a) Hard code 64MB.
b) Use 12.8% of the `apm_config.max_memory` value (which is 64MB by default).
c) Add an undocumented (?) configuration knob.

My vote stands for b) to avoid having even more knobs and to make it proportional to what the user wants to allocate to the process.